src/Stratis.SmartContracts.Core/SmartContractScript.cs

```diff
@@ -1,43 +1,62 @@
 ﻿using NBitcoin;
+using TracerAttributes;
 
 namespace Stratis.SmartContracts.Core
 {
     public static class SmartContractScript
     {
+        [NoTrace]
         public static bool IsSmartContractExec(this Script script)
         {
-            return script.IsSmartContractCall() || script.IsSmartContractCreate();
+            return TestFirstByte(script, new byte[]{ (byte)ScOpcodeType.OP_CALLCONTRACT, (byte)ScOpcodeType.OP_CREATECONTRACT });
         }
 
+        [NoTrace]
         public static bool IsSmartContractCall(this Script script)
         {
-            return TestFirstByte(script, (byte)ScOpcodeType.OP_CALLCONTRACT);
+            return TestFirstByte(script,  new byte[]{(byte)ScOpcodeType.OP_CALLCONTRACT });
         }
 
+        [NoTrace]
         public static bool IsSmartContractCreate(this Script script)
         {
-            return TestFirstByte(script, (byte)ScOpcodeType.OP_CREATECONTRACT);
+            return TestFirstByte(script, new byte[] {(byte)ScOpcodeType.OP_CREATECONTRACT });
         }
 
+        [NoTrace]
         public static bool IsSmartContractSpend(this Script script)
         {
-            return TestFirstByte(script, (byte)ScOpcodeType.OP_SPEND);
-
+            return TestFirstByte(script, new byte[] {(byte)ScOpcodeType.OP_SPEND });
         }
 
+        [NoTrace]
         public static bool IsSmartContractInternalCall(this Script script)
         {
-            return TestFirstByte(script, (byte) ScOpcodeType.OP_INTERNALCONTRACTTRANSFER);
+            return TestFirstByte(script, new byte[]{(byte) ScOpcodeType.OP_INTERNALCONTRACTTRANSFER });
         }
 
-        private static bool TestFirstByte(Script script, byte opcode)
+        /// <summary>
+        /// Tests whether a script starts with any of the given opcodes.
+        /// </summary>
+        /// <param name="script">The script to test.</param>
+        /// <param name="opcodes">A list of opcodes in byte form to test against.</param>
+        /// <returns>True if the script starts with one of the given opcodes.</returns>
+        [NoTrace]
+        private static bool TestFirstByte(Script script, byte[] opcodes)
```

Comment:
Don't like using an array of opcodes here - this should just be a simple test. Code is harder to understand and negligible performance benefit. Plus an array allocation every time you call it with a single opcode.

Reply:
Ah yes, it was worse when it was doing tobytes every time but in this case it makes sense

Reply:
We could also pass the script bytes in and avoid the duplicate call to `ToBytes`.

Reply:
The unsafe ToBytes just returns the bytes array inside of the Script so it's chill